---
title: "Untitled"
output: pdf_document
date: "2024-06-14"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r set values}
thresh = 0.2
colour.amplification = "red3"
colour.loss = "blue4"
array_type = "450k"
ArrayType = "450k"
mSetsAnno <- list("anno_targets" = anno_targets)
```


```{r singleFrequencyPlot}
#' Generates the single sample segmentation aberration frequency plot
#' @param mSetsAnno A list of the RGSet of the target data, the control data and the annotation data
#' @param segmentation_data a dataframe of the segmentation results of the single sample segmentation.
#' @param target_ratios a dataframe of the intesity of all bins for each sample.
#' @param colour.amplification Colour for amplification
#' @param colour.loss Colour for loss
#' @param thresh a float specifying the threshold when a segment is called as an aberration
#'
#' @return returns the frequency plot of the samples that were segmented together using the single sample segmentation algorithm
singleFrequencyPlot <- function(mSetsAnno, segmentation_data, colour.amplification, colour.loss, thresh, array_type){

    #genome data
    chr_name <- mSetsAnno$anno_targets@genome$chr
    chr_size <- mSetsAnno$anno_targets@genome$size
    genome_chr <- cumsum(as.numeric(chr_size))
    chr_centr <- mSetsAnno$anno_targets@genome$pq
    
    #just for getting pq - lines
    addChr <- c(0,genome_chr)
    addChr <- addChr[-length(addChr)]
    genome_centr <- chr_centr + addChr
    
    axis_break <- genome_centr
    axis_label <- c(1:22)
    
    cnFreq_seg <- as.data.frame(GenVisR::cnFreq(segmentation_data, CN_low_cutoff = -thresh,  CN_high_cutoff = thresh, out = "data"))
    cnFreq_seg$data.chromosome <- gsub("chr", "", cnFreq_seg$data.chromosome)
    cnFreq_seg$data.chromosome <- as.numeric(cnFreq_seg$data.chromosome)
    
    data.start = c()
    data.end = c()
    counterCnFreq <- 1:length(cnFreq_seg$data.chromosome)
    for(i in counterCnFreq){
      data.start[i] <- cnFreq_seg$data.start[i] + addChr[cnFreq_seg$data.chromosome[i]]
      data.end[i] <- cnFreq_seg$data.end[i] + addChr[cnFreq_seg$data.chromosome[i]]
    }
    
    cnFreq_seg$data.start <- as.numeric(data.start)
    cnFreq_seg$data.end <- as.numeric(data.end)
    cnFreq_seg$data.lossProportion <- cnFreq_seg$data.lossProportion*-1
    
    #duplicate rows/segment with gains and loss
    df.expanded <- cnFreq_seg[rep(row.names(cnFreq_seg), ifelse(cnFreq_seg$data.gainProportion != 0 & cnFreq_seg$data.lossProportion != 0 ,2 ,1)),]
    
    #now every row is one rectangle of the final plot, extra rows have .1 in their name
    df.expanded$new_rownames <- rownames(df.expanded)
    
    #define the heights of these rectangles
    df.expanded$Y_MIN <- ifelse(df.expanded$data.gainProportion == 0 | df.expanded$data.lossProportion == 0, df.expanded$data.lossProportion, ifelse(grepl("\\.1", df.expanded$new_rownames), df.expanded$data.lossProportion, 0))
    
    df.expanded$Y_MAX <- ifelse(df.expanded$data.gainProportion == 0 | df.expanded$data.lossProportion == 0, df.expanded$data.gainProportion, ifelse(grepl("\\.1", df.expanded$new_rownames), 0, df.expanded$data.gainProportion))
    
    #decribe segment properties
    df.expanded$type <- ifelse(df.expanded$data.gainProportion == 0 & df.expanded$data.lossProportion == 0, "neutral", ifelse(df.expanded$Y_MAX == 0, "loss", "gain"))
    
    #separate into two dataframes for filtering
    gains <- df.expanded[which(df.expanded$type == "gain"),]
    loss <- df.expanded[which(df.expanded$type == "loss"),]
    
    gains_filtered_for_plot <- as.data.frame(cbind(gains$data.start, gains$data.end, gains$data.gainProportion, gains$type))
    loss_filtered_for_plot <- as.data.frame(cbind(loss$data.start, loss$data.end, loss$data.lossProportion, loss$type))
    
    #names(gains_filtered_for_plot) <- c("start", "end", "count", "type")
    names(loss_filtered_for_plot) <- names(gains_filtered_for_plot)
    
    to_plot <- as.data.frame(rbind(gains_filtered_for_plot, loss_filtered_for_plot ))
    names(to_plot) <- c("start", "end", "count", "type")
    
    y_axis_break <-c(-1,-0.75, -0.5, -0.25, 0, 0.25, 0.5, 0.75, 1)
    y_axis_label <- c("100","75", "50","25", "0","25", "50","75", "100" )
    
    to_plot$start <- as.numeric(to_plot$start)
    to_plot$end <- as.numeric(to_plot$end)
    for(i in 1:length(to_plot$end)){
      if ((to_plot$end[i] - to_plot$start[i]) <5000000) {
        to_plot$end[i] <- to_plot$start[i] + 5000000
      }
    }
    
    to_plot$count <- as.numeric(to_plot$count)
    
    singleFreqPlot <- ggplot2::ggplot() + geom_rect(data = to_plot,  aes(xmin = start, xmax = end, ymax = count, ymin = 0 , fill = type))  +
      ggplot2::geom_vline(xintercept = genome_chr, colour = "grey") + 
      ggplot2::geom_vline(xintercept = genome_centr, linetype="dotted", colour = "grey") + 
      ggplot2::geom_hline(yintercept = 0, colour ="darkgrey") + 
      ggplot2::scale_x_continuous(name = "Chromosome", breaks = axis_break, labels = axis_label, limits = c(0, 2881033286)) +
      ggplot2::theme_classic(base_size = 12) + 
      ggplot2::guides(x = guide_axis(angle = 40)) +
      ggplot2::scale_y_continuous(name = "Occurence [%]", breaks = y_axis_break, labels = y_axis_label, limits = c(-1, 1))+
      ggplot2::scale_fill_discrete(guide="none") + 
      ggplot2::scale_fill_manual(values = c("gain" = colour.amplification,"loss" = colour.loss), guide="none") + 
      ggplot2::coord_cartesian(xlim = c(50001, 2881033286), expand = FALSE) + 
      ggplot2::guides(x = guide_axis(n.dodge = 2))
 
  
  return(singleFreqPlot)
}
```

```{r overlayPlot}
#' Generates the single sample segmentation aberration frequency plot
#' @param mSetsAnno A list of the RGSet of the target data, the control data and the annotation data
#' @param segmentation_data a dataframe of the segmentation results of the single sample segmentation.
#' @param colour.amplification Colour for amplification
#' @param colour.loss Colour for loss
#'
#' @return returns an overlay plot of all segmentations resulting from the single sample segmentation algorithm
overlayPlot <- function(mSetsAnno, segmentation_data, colour.amplification, colour.loss , array_type) {
  if (array_type == "mouse") {
    #genome data
    chr_name <- mSetsAnno$anno_targets@genome$chr
    chr_size <- mSetsAnno$anno_targets@genome$size
    genome_chr <- cumsum(as.numeric(chr_size))
    
    #just for getting pq - lines
    addChr <- c(0,genome_chr)
    addChr <- addChr[-length(addChr)]
    
    #calculate position on genome
    seg_start <- c()
    seg_end <- c()
    foreach(i=1:length(segmentation_data$chromosome)) %do%
      {
        seg_start[i] <- segmentation_data$start[i] + addChr[segmentation_data$chromosome[i]]
        seg_end[i] <- segmentation_data$end[i] + addChr[segmentation_data$chromosome[i]] 
        if ((segmentation_data$end[i] - segmentation_data$start[i]) <5000000) {
          segmentation_data$end[i] <- segmentation_data$start[i] + 5000000
        }
      }
    
    segmentation_data$start <- seg_start
    segmentation_data$end <- seg_end
    sample_no <- as.numeric(length(unique(segmentation_data$sample)))
    
    axis_break <- genome_chr
    axis_label <- c(1:length(genome_chr))
    
    segcount <- c(1:length(segmentation_data$segmean))
    segmentation_data$col <- "amplification"
    for (i in segcount) {
      if (segmentation_data$segmean[i] < 0)
      {
        segmentation_data$col[i] <- "deletion"
      }
    }
    
    overlayPlot <- ggplot2::ggplot() + 
      ggplot2::geom_rect(data = segmentation_data,  aes(xmin = start, xmax = end, ymax = segmean, ymin = 0 , fill = col))  +
      ggplot2::geom_vline(xintercept = genome_chr, colour = "grey") +
      ggplot2::ylim(-2.5, 2.5) + 
      ggplot2::geom_hline( yintercept = 0, colour ="darkgrey") +
      ggplot2::scale_x_continuous(name = "Chromosome", breaks = axis_break, labels = axis_label, limits = c(0, 2881033286)) +
      ggplot2::theme_classic(base_size = 15) +
      ggplot2::guides(x = guide_axis(angle = 40)) +
      ggplot2::scale_fill_discrete(guide="none") +
      ggplot2::coord_cartesian(xlim = c(50001, 2881033286), expand = FALSE) +
      ggplot2::guides(x = guide_axis(n.dodge = 2))+
      ggplot2::scale_fill_manual(values = alpha(c("amplification" = colour.amplification,"deletion" = colour.loss), 1/sample_no), guide="none")
    print(overlayPlot)
  }
  else {
    #genome data
    chr_name <- mSetsAnno$anno_targets@genome$chr
    chr_size <- mSetsAnno$anno_targets@genome$size
    genome_chr <- cumsum(as.numeric(chr_size))
    chr_centr <- mSetsAnno$anno_targets@genome$pq
    
    #just for getting pq - lines
    addChr <- c(0,genome_chr)
    addChr <- addChr[-length(addChr)]
    genome_centr <- chr_centr + addChr
    
    #calculate position on genome
    seg_start <- c()
    seg_end <- c()
    foreach(i=1:length(segmentation_data$chromosome)) %do%
      {
        seg_start[i] <- segmentation_data$start[i] + addChr[segmentation_data$chromosome[i]]
        seg_end[i] <- segmentation_data$end[i] + addChr[segmentation_data$chromosome[i]] 
        if ((segmentation_data$end[i] - segmentation_data$start[i]) <5000000) {
          segmentation_data$end[i] <- segmentation_data$start[i] + 5000000
        }
      }
    
    segmentation_data$start <- seg_start
    segmentation_data$end <- seg_end
    sample_no <- as.numeric(length(unique(segmentation_data$sample)))
    
    axis_break <- genome_centr
    axis_label <- c(1:length(genome_centr))
    
    segcount <- c(1:length(segmentation_data$segmean))
    segmentation_data$col <- "amplification"
    for (i in segcount) {
      if (segmentation_data$segmean[i] < 0)
      {
        segmentation_data$col[i] <- "deletion"
      }
    }
    
    overlayPlot <- ggplot2::ggplot() + 
      ggplot2::geom_rect(data = segmentation_data,  aes(xmin = start, xmax = end, ymax = segmean, ymin = 0 , fill = col))  +
      ggplot2::geom_vline(xintercept = genome_chr, colour = "grey") +
      ggplot2::ylim(-2.5, 2.5) + 
      ggplot2::geom_vline(xintercept = genome_centr, linetype="dotted", colour = "grey") +
      ggplot2::geom_hline( yintercept = 0, colour ="darkgrey") +
      ggplot2::scale_x_continuous(name = "Chromosome", breaks = axis_break, labels = axis_label, limits = c(0, 2881033286)) +
      ggplot2::theme_classic(base_size = 15) +
      ggplot2::guides(x = guide_axis(angle = 40)) +
      ggplot2::scale_fill_discrete(guide="none") +
      ggplot2::coord_cartesian(xlim = c(50001, 2881033286), expand = FALSE) +
      ggplot2::guides(x = guide_axis(n.dodge = 2))+
      ggplot2::scale_fill_manual(values = alpha(c("amplification" = colour.amplification,"deletion" = colour.loss), 1/sample_no), guide="none")
    
  }

  return(overlayPlot)
  
}
```

```{r CCNV}
#' Generates the cumulative copy number plot
#' @param mSetsAnno A list of the RGSet of the target data, the control data and the annotation data
#' @param seg_mpcf a dataframe of the segmentation results of the multi sample segmentation.
#' @param target_ratios a dataframe of the intesity of all bins for each sample.
#' @param colour.amplification Colour for amplification
#' @param colour.loss Colour for loss
#' @param detail.regions Either NULL or a vector of gene names.
#'
#' @return returns the plot of the cumulative copy number variation plot

CCNV <- function(mSetsAnno, seg_mpcf, target_ratios, ArrayType, colour.amplification, colour.loss, detail.regions, array_type){
  annotate = TRUE
  if(is.null(detail.regions)){
    annotate = FALSE
  }
  
  if (array_type == "mouse") {
    bin_data_samples <- target_ratios[,-(which(names(target_ratios) %in% c("Chrom", "Median.bp")))]
    
    #here anno longer due to na.omit
    bin_data_all <- as.data.frame(cbind(as.numeric(mSetsAnno$anno_targets@bins@ranges@start), as.numeric(rowMeans(bin_data_samples)), target_ratios$Chrom))
    colnames(bin_data_all)<- c("bin_start", "bin_value", "names")
    
    #genome data
    chr_name <- mSetsAnno$anno_targets@genome$chr
    chr_size <- mSetsAnno$anno_targets@genome$size
    genome_chr <- cumsum(as.numeric(chr_size))
    
    #just for getting pq - lines
    addChr <- c(0,genome_chr)
    addChr <- addChr[-length(addChr)]
    
    #calculate position on genome
    start = c()
    foreach(i=1:length(bin_data_all$bin_start)) %do%
      {
        start[i] <- bin_data_all$bin_start[i] + addChr[bin_data_all$names[i]]
      }
    
    # bin data consisting out of the starting point of the bin and the value
    final_bin_data <- as.data.frame(cbind(as.numeric(start), bin_data_all$bin_value))
    final_bin_data$bin_nr <- rownames(final_bin_data)
    
    #segment data
    seg_data_samples <- seg_mpcf[,-(which(names(seg_mpcf) %in% c("chrom", "arm", "start.pos", "end.pos", "n.probes")))]
    #mean seg value
    segSta <- as.data.frame(cbind(seg_mpcf$start.pos, rowMeans(seg_data_samples), as.numeric(seg_mpcf$chrom)))
    segEnd <- as.data.frame(cbind(seg_mpcf$end.pos, rowMeans(seg_data_samples), as.numeric(seg_mpcf$chrom)))
    names(segEnd) <- names(segSta)
    segment_data <- as.data.frame(rbind(segSta, segEnd))
    colnames(segment_data) <- c("seg_lim", "seg_val", "chr")
    
    #calculate position on genome
    seg_start = c()
    foreach(i=1:length(segment_data$chr)) %do%
      {
        seg_start[i] <- segment_data$seg_lim[i] + addChr[segment_data$chr[i]]
      }
    
    final_seg_data <- as.data.frame(cbind(as.numeric(seg_start), as.numeric(segment_data$seg_val)))
    final_seg_data <-final_seg_data[order(final_seg_data$V1),]
    
    #x-axis
    axis_break <- genome_chr
    axis_label <- c(1:length(genome_chr))
    b <- c(-1, -0.5, 0, 0.5, 1)
    
    
    if (annotate == TRUE) {
      
      print("Annotation function is not yet accessible for mouse data")
      
      cumCNV <- ggplot2::ggplot() +
        ggplot2::geom_point(final_bin_data, mapping = aes(x=V1, y=V2, colour=V2))+ 
        ggplot2::ylim(-2.5, 2.5) + 
        ggplot2::scale_color_gradient2(low = colour.loss ,  mid = "white",  high = colour.amplification,  midpoint = 0, breaks = b, guide = "colourbar" , space = "Lab", name = "Intensity") + 
        ggplot2::geom_vline(xintercept = genome_chr, colour = "grey") + 
        ggplot2::geom_path(data = final_seg_data, aes(x = V1, y = V2)) + 
        ggplot2::labs(y = "Intensity") +
        ggplot2::scale_x_continuous(limits = c(10001, 2881033286), name = "Chromosome", breaks = axis_break, labels = axis_label ) +
        ggplot2::theme_classic(base_size = 15) + 
        ggplot2::coord_cartesian(xlim = c(50001, 2881033286), expand = FALSE) + 
        ggplot2::guides(x = guide_axis(n.dodge = 2))
    }
    
    if (annotate == FALSE){
      cumCNV <- ggplot2::ggplot() +
        ggplot2::geom_point(final_bin_data, mapping = aes(x=V1, y=V2, colour=V2))+ 
        ggplot2::ylim(-2.5, 2.5) + 
        ggplot2::scale_color_gradient2(low = colour.loss ,  mid = "white",  high = colour.amplification,  midpoint = 0, breaks = b, guide = "colourbar" , space = "Lab", name = "Intensity") + 
        ggplot2::geom_vline(xintercept = genome_chr, colour = "grey") + 
        ggplot2::geom_path(data = final_seg_data, aes(x = V1, y = V2)) + 
        ggplot2::labs(y = "Intensity") +
        ggplot2::scale_x_continuous(limits = c(10001, 2881033286), name = "Chromosome", breaks = axis_break, labels = axis_label ) +
        ggplot2::theme_classic(base_size = 15) + 
        ggplot2::coord_cartesian(xlim = c(50001, 2881033286), expand = FALSE) + 
        ggplot2::guides(x = guide_axis(n.dodge = 2))
    }
  }
  else {
    bin_data_samples <- target_ratios[,-(which(names(target_ratios) %in% c("Chrom", "Median.bp")))]
    bin_data_all <- as.data.frame(cbind(as.numeric(mSetsAnno$anno_targets@bins@ranges@start), as.numeric(rowMeans(bin_data_samples)), target_ratios$Chrom))
    colnames(bin_data_all)<- c("bin_start", "bin_value", "names")
    
    #genome data
    chr_name <- mSetsAnno$anno_targets@genome$chr
    chr_size <- mSetsAnno$anno_targets@genome$size
    genome_chr <- cumsum(as.numeric(chr_size))
    chr_centr <- mSetsAnno$anno_targets@genome$pq
    
    #just for getting pq - lines
    addChr <- c(0,genome_chr)
    addChr <- addChr[-length(addChr)]
    genome_centr <- chr_centr + addChr
    
    #calculate position on genome
    start = c()
    foreach(i=1:length(bin_data_all$bin_start)) %do%
      {
        start[i] <- bin_data_all$bin_start[i] + addChr[bin_data_all$names[i]]
      }
    
    # bin data consisting out of the starting point of the bin and the value
    final_bin_data <- as.data.frame(cbind(as.numeric(start), bin_data_all$bin_value))
    final_bin_data$bin_nr <- rownames(final_bin_data)
    
    #segment data
    seg_data_samples <- seg_mpcf[,-(which(names(seg_mpcf) %in% c("chrom", "arm", "start.pos", "end.pos", "n.probes")))]
    #mean seg value
    segSta <- as.data.frame(cbind(seg_mpcf$start.pos, rowMeans(seg_data_samples), as.numeric(seg_mpcf$chrom)))
    segEnd <- as.data.frame(cbind(seg_mpcf$end.pos, rowMeans(seg_data_samples), as.numeric(seg_mpcf$chrom)))
    names(segEnd) <- names(segSta)
    segment_data <- as.data.frame(rbind(segSta, segEnd))
    colnames(segment_data) <- c("seg_lim", "seg_val", "chr")
    
    #calculate position on genome
    seg_start = c()
    foreach(i=1:length(segment_data$chr)) %do%
      {
        seg_start[i] <- segment_data$seg_lim[i] + addChr[segment_data$chr[i]]
      }
    
    final_seg_data <- as.data.frame(cbind(as.numeric(seg_start), as.numeric(segment_data$seg_val)))
    final_seg_data <-final_seg_data[order(final_seg_data$V1),]
    
    #x-axis
    axis_break <- genome_centr
    axis_label <- c(1:22)
    b <- c(-1, -0.5, 0, 0.5, 1)
    
    
    if (annotate == TRUE) {
      if (ArrayType != "450k"){
        anno <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
      }
      if (ArrayType == "EPIC"){
        anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
      }
      if (ArrayType == "EPICv2"){
        print("Annotation function is not yet accessible for EPICv2 Arrays")
        cumCNV <- ggplot2::ggplot() +
          ggplot2::geom_point(final_bin_data, mapping = aes(x=V1, y=V2, colour=V2))+ 
          ggplot2::ylim(-2.5, 2.5) + 
          ggplot2::scale_color_gradient2(low = colour.loss ,  mid = "white",  high = colour.amplification,  midpoint = 0, breaks = b, guide = "colourbar" , space = "Lab", name = "Intensity") + 
          ggplot2::geom_vline(xintercept = genome_chr, colour = "grey") + 
          ggplot2::geom_vline(xintercept = genome_centr,linetype="dotted", colour = "grey")  + 
          ggplot2::geom_path(data = final_seg_data, aes(x = V1, y = V2)) + 
          ggplot2::labs(y = "Intensity") +
          ggplot2::scale_x_continuous(limits = c(10001, 2881033286), name = "Chromosome", breaks = axis_break, labels = axis_label ) +
          ggplot2::theme_classic(base_size = 15) + 
          ggplot2::coord_cartesian(xlim = c(50001, 2881033286), expand = FALSE) + 
          ggplot2::guides(x = guide_axis(n.dodge = 2))
      }
      
      #detail data -> here used to be anno _EPIC
      gene_regions <- as.data.frame(cbind(anno@listData$UCSC_RefGene_Name, anno@rownames, anno@listData$chr, anno@listData$pos))
      colnames(gene_regions) <- c("genes", "CpG_sites", "chr", "pos")
      gene_regions$genes <- gsub(";"," ",gene_regions$genes)
      gene_regions$chr <- gsub("chr", "", gene_regions$chr)
      df_gene_regions <- gene_regions[!(gene_regions$chr=="X" | gene_regions$chr=="Y"),]
      df_gene_regions$chr <- as.numeric(df_gene_regions$chr)
      df_gene_regions$pos <- as.numeric(df_gene_regions$pos)
      
      #calculate position on genome
      cpg_pos = c()
      counter3 <- 1:length(df_gene_regions$chr)
      for(i in counter3){
        cpg_pos[i] <- df_gene_regions$pos[i] + addChr[df_gene_regions$chr[i]]
      }
      df_gene_regions$pos <- as.numeric(cpg_pos)
      genes <- as.data.frame(detail.regions)
      
      final_bin_data <-  na.omit(final_bin_data)
      df_gene_regions$bin <- final_bin_data$bin_nr[findInterval(df_gene_regions$pos, final_bin_data$V1)]
      df_gene_regions <- as.data.frame(na.omit(as.matrix(df_gene_regions)))
      df_gene_regions <- tidyr::separate(df_gene_regions, genes, into = c("genes1", "genes2"), sep = " (?=[^ ]+$)")
      
      #put together the genes and the corresponding bin data
      geneBins <- as.data.frame(dplyr::left_join(genes, df_gene_regions, by = dplyr::join_by("detail.regions" == "genes1")))
      
      #remove duplicates and only keep the bin with the highest number of CpG sites present
      geneBins <- geneBins %>% dplyr::group_by(geneBins$bins) %>% dplyr::count(geneBins$detail.regions, geneBins$bin)
      colnames(geneBins) <- c("genes", "bins", "n")
      geneBins <- geneBins[order(geneBins$n, decreasing  = TRUE),]
      geneBins <- na.omit(geneBins[-which(duplicated(geneBins$genes)),])
      
      geneBins <- dplyr::left_join(geneBins, final_bin_data, by= dplyr::join_by("bins" == "bin_nr"))
      geneBins <- geneBins[!((abs(geneBins$V2)) <0.15),]
      
      cumCNV <- ggplot2::ggplot() +
        ggplot2::geom_point(final_bin_data, mapping = aes(x=V1, y=V2, colour=V2))+ 
        ggplot2::ylim(-2.5, 2.5) + 
        ggplot2::scale_color_gradient2(low = colour.loss ,  mid = "white",  high = colour.amplification,  midpoint = 0, breaks = b, guide = "colourbar" , space = "Lab", name = "Intensity") + 
        ggplot2::geom_vline(xintercept = genome_chr, colour = "grey") + 
        ggplot2::geom_vline(xintercept = genome_centr,linetype="dotted", colour = "grey")  + 
        ggplot2::geom_path(data = final_seg_data, aes(x = V1, y = V2)) + 
        ggplot2::labs( y = "Intensity") + 
        ggplot2::geom_point(data = geneBins, aes(x = V1, y = V2)) + 
        ggrepel::geom_label_repel(data = geneBins, aes(x = V1, y = V2), label = geneBins$genes, label.size = 0.2, nudge_x = 0.1, nudge_y = 0.1,) + 
        ggplot2::scale_x_continuous(limits = c(10001, 2881033286), name = "Chromosome", breaks = axis_break, labels = axis_label ) +
        ggplot2::theme_classic(base_size = 15) + 
        ggplot2::coord_cartesian(xlim = c(50001, 2881033286), expand = FALSE) + 
        ggplot2::guides(x = guide_axis(n.dodge = 2))
    }
    
    if (annotate == FALSE){
      cumCNV <- ggplot2::ggplot() +
        ggplot2::geom_point(final_bin_data, mapping = aes(x=V1, y=V2, colour=V2))+ 
        ggplot2::ylim(-2.5, 2.5) + 
        ggplot2::scale_color_gradient2(low = colour.loss ,  mid = "white",  high = colour.amplification,  midpoint = 0, breaks = b, guide = "colourbar" , space = "Lab", name = "Intensity") + 
        ggplot2::geom_vline(xintercept = genome_chr, colour = "grey") + 
        ggplot2::geom_vline(xintercept = genome_centr,linetype="dotted", colour = "grey")  + 
        ggplot2::geom_path(data = final_seg_data, aes(x = V1, y = V2)) + 
        ggplot2::labs(y = "Intensity") +
        ggplot2::scale_x_continuous(limits = c(10001, 2881033286), name = "Chromosome", breaks = axis_break, labels = axis_label ) +
        ggplot2::theme_classic(base_size = 15) + 
        ggplot2::coord_cartesian(xlim = c(50001, 2881033286), expand = FALSE) + 
        ggplot2::guides(x = guide_axis(n.dodge = 2))
    }
  }
  
  return(cumCNV)
}
```

```{r cumFreq}
#' Generates the multi sample segmentation aberration frequency plot
#' @param mSetsAnno A list of the RGSet of the target data, the control data and the annotation data
#' @param seg_mpcf a dataframe of the segmentation results of the multi sample segmentation.
#' @param target_ratios a dataframe of the intesity of all bins for each sample.
#' @param colour.amplification Colour for amplification
#' @param colour.loss Colour for loss
#' @param thresh a float specifying the threshold when a segment is called as an aberration
#'
#' @return returns the frequency plot of the samples that were segmented togethher using the multi sample segmentation algorithm
#' 
cumFreq <- function(mSetsAnno, seg_mpcf, target_ratios, colour.amplification, colour.loss, thresh, array_type){
  if (array_type == "mouse") {
    #Genome Data
    chr_name <- mSetsAnno$anno_targets@genome$chr
    chr_size <- mSetsAnno$anno_targets@genome$size
    genome_chr <- cumsum(as.numeric(chr_size))
    
    #just for getting pq - lines
    addChr <- c(0,genome_chr)
    addChr <- addChr[-length(addChr)]
    
    #defining threshholds
    broad_min_probes <- 300
    focal_min_probes <- 2
    
    #x-axis
    axis_break <- genome_chr
    axis_label <- c(1:length(genome_chr))
    b <- c(-1, -0.5, 0, 0.5, 1)
    
    #segment
    segstart <- seg_mpcf$start.pos
    segend <- seg_mpcf$end.pos
    segnames <- as.numeric(seg_mpcf$chrom)
    
    segment_data <- as.data.frame(cbind(segstart, segend, segnames))
    colnames(segment_data) <- c("seg_start", "seg_end" , "chr")
    
    #calculate position on genome
    seg_start <- c()
    seg_end <- c()
    foreach(i=1:length(segment_data$chr)) %do%
      {
        seg_start[i] <- segment_data$seg_start[i] + addChr[segment_data$chr[i]]
        seg_end[i] <- segment_data$seg_end[i] + addChr[segment_data$chr[i]]
      }
    
    seg_data_samples <- seg_mpcf[,-(which(names(seg_mpcf) %in% c("chrom", "arm", "start.pos", "end.pos", "n.probes")))]
    df_seg_val <- as.data.frame(cbind(seg_start, seg_end,  rowMeans(seg_data_samples)))
    df_abberation_candidates <- as.data.frame(cbind(seg_start, seg_end, seg_mpcf$n.probes, seg_data_samples))
    
    #focal candidates
    focal_candidates <- df_abberation_candidates
    #focal_candidates$seg_end <- focal_candidates$seg_end + 10000000
    for (i in 1:length(focal_candidates$seg_end)) {
      if ((focal_candidates$seg_end[i] - focal_candidates$seg_start[i]) <5000000) {
        focal_candidates$seg_end[i] <- focal_candidates$seg_start[i] + 5000000
      }
    }
    focal_candidates <- focal_candidates[!((abs(focal_candidates$`seg_mpcf$n.probes`)) < focal_min_probes),]
    focal_candidates$mean <- rowMeans(seg_data_samples)
    focal_candidates$col <- "amplification"
    focal_candid_values <- focal_candidates[,-(1:5)]
    focalcount <- c(1:length(focal_candidates$seg_start))
    for (i in focalcount) {
      focal_candidates$count[i] <- max(0, (sum(focal_candid_values[i,] >= thresh)) )
      if (focal_candidates$mean[i] < 0)
      {
        focal_candidates$count[i] <- focal_candidates$count[i] * -1
        focal_candidates$col[i] <- "deletion"
      }
    }
    
    #broad candidates
    broad_candidates <- df_abberation_candidates
    broad_candidates$mean <- rowMeans(seg_data_samples)
    broad_candidates <- broad_candidates[!((abs(broad_candidates$`seg_mpcf$n.probes`)) < broad_min_probes),]
    broad_candid_values <- abs(broad_candidates[,-(1:5)])
    broadcount <- c(1:length(broad_candid_values$mean))
    broad_candidates$col <- "amplification"
    for (i in broadcount) {
      broad_candidates$count[i] <- max(0, (sum(broad_candid_values[i,] >= thresh)) )
      if (broad_candidates$mean[i] < 0)
      {
        broad_candidates$count[i] <- broad_candidates$count[i] * -1
        broad_candidates$col[i] <- "deletion"
      }
    }
    if (nrow(focal_candidates[(focal_candidates$count!=0),]) == 0) {
      candidates <- broad_candidates[(broad_candidates$count!=0),]
    } else if (nrow(broad_candidates[(broad_candidates$count!=0),]) == 0) {
      candidates <- focal_candidates[(focal_candidates$count!=0),]
    } else {
      candidates <- rbind(focal_candidates[(focal_candidates$count!=0),], broad_candidates[(broad_candidates$count!=0),])
    }
    sample_no <- ncol(seg_data_samples)
    b <- c(-1, -0.5, 0, 0.5, 1)
    y_axis_break <-c(-sample_no,-(sample_no/2) + -(sample_no/4), -(sample_no/2), -(sample_no/4),0, sample_no/4, sample_no/2,(sample_no/2) + (sample_no/4), sample_no)
    y_axis_label <- c("100","75", "50","25", "0","25", "50","75", "100" )
    
    cumFreq <- ggplot2::ggplot() + 
      ggplot2::geom_rect(data = candidates,  aes(xmin = seg_start, xmax = seg_end, ymax = count, ymin = 0 , fill = col))  +
      ggplot2::geom_vline(xintercept = genome_chr, colour = "grey") +
      ggplot2::geom_hline( yintercept = 0, colour ="darkgrey") +
      ggplot2::scale_x_continuous(name = "Chromosome", breaks = axis_break, labels = axis_label, limits = c(0, 2881033286)) +
      ggplot2::theme_classic(base_size = 15) +
      ggplot2::guides(x = guide_axis(angle = 40)) +
      ggplot2::scale_y_continuous(name = "Occurence [%]", breaks = y_axis_break, labels = y_axis_label, limits = c(-sample_no, sample_no)) +
      ggplot2::scale_fill_discrete(guide="none") +
      ggplot2::scale_fill_manual(values = c("amplification" = colour.amplification,"deletion" = colour.loss), guide="none") +
      ggplot2::coord_cartesian(xlim = c(50001, 2881033286), expand = FALSE) +
      ggplot2::guides(x = guide_axis(n.dodge = 2))
  }
  else {
    chr_name <- mSetsAnno$anno_targets@genome$chr
    chr_size <- mSetsAnno$anno_targets@genome$size
    genome_chr <- cumsum(as.numeric(chr_size))
    chr_centr <- mSetsAnno$anno_targets@genome$pq
    
    #just for getting pq - lines
    addChr <- c(0,genome_chr)
    addChr <- addChr[-length(addChr)]
    genome_centr <- chr_centr + addChr
    
    #defining threshholds
    broad_min_probes <- 300
    focal_min_probes <- 2
    
    #x-axis
    axis_break <- genome_centr
    axis_label <- c(1:22)
    b <- c(-1, -0.5, 0, 0.5, 1)
    
    #segment
    segstart <- seg_mpcf$start.pos
    segend <- seg_mpcf$end.pos
    segnames <- as.numeric(seg_mpcf$chrom)
    
    segment_data <- as.data.frame(cbind(segstart, segend, segnames))
    colnames(segment_data) <- c("seg_start", "seg_end" , "chr")
    
    #calculate position on genome
    seg_start <- c()
    seg_end <- c()
    foreach(i=1:length(segment_data$chr)) %do%
      {
        seg_start[i] <- segment_data$seg_start[i] + addChr[segment_data$chr[i]]
        seg_end[i] <- segment_data$seg_end[i] + addChr[segment_data$chr[i]]
      }
    
    seg_data_samples <- seg_mpcf[,-(which(names(seg_mpcf) %in% c("chrom", "arm", "start.pos", "end.pos", "n.probes")))]
    df_seg_val <- as.data.frame(cbind(seg_start, seg_end,  rowMeans(seg_data_samples)))
    df_abberation_candidates <- as.data.frame(cbind(seg_start, seg_end, seg_mpcf$n.probes, seg_data_samples))
    
    #focal candidates
    focal_candidates <- df_abberation_candidates
    #focal_candidates$seg_end <- focal_candidates$seg_end + 10000000
    for (i in 1:length(focal_candidates$seg_end)) {
      if ((focal_candidates$seg_end[i] - focal_candidates$seg_start[i]) <5000000) {
        focal_candidates$seg_end[i] <- focal_candidates$seg_start[i] + 5000000
      }
    }
    focal_candidates <- focal_candidates[!((abs(focal_candidates$`seg_mpcf$n.probes`)) < focal_min_probes),]
    focal_candidates$mean <- rowMeans(seg_data_samples)
    focal_candidates$col <- "amplification"
    focal_candid_values <- focal_candidates[,-(1:5)]
    focalcount <- c(1:length(focal_candidates$seg_start))
    for (i in focalcount) {
      focal_candidates$count[i] <- max(0, (sum(focal_candid_values[i,] >= thresh)) )
      if (focal_candidates$mean[i] < 0)
      {
        focal_candidates$count[i] <- focal_candidates$count[i] * -1
        focal_candidates$col[i] <- "deletion"
      }
    }
    
    #broad candidates
    broad_candidates <- df_abberation_candidates
    broad_candidates$mean <- rowMeans(seg_data_samples)
    broad_candidates <- broad_candidates[!((abs(broad_candidates$`seg_mpcf$n.probes`)) < broad_min_probes),]
    broad_candid_values <- abs(broad_candidates[,-(1:5)])
    broadcount <- c(1:length(broad_candid_values$mean))
    broad_candidates$col <- "amplification"
    for (i in broadcount) {
      broad_candidates$count[i] <- max(0, (sum(broad_candid_values[i,] >= thresh)) )
      if (broad_candidates$mean[i] < 0)
      {
        broad_candidates$count[i] <- broad_candidates$count[i] * -1
        broad_candidates$col[i] <- "deletion"
      }
    }
    if (nrow(focal_candidates[(focal_candidates$count!=0),]) == 0) {
      candidates <- broad_candidates[(broad_candidates$count!=0),]
    } else if (nrow(broad_candidates[(broad_candidates$count!=0),]) == 0) {
      candidates <- focal_candidates[(focal_candidates$count!=0),]
    } else {
      candidates <- rbind(focal_candidates[(focal_candidates$count!=0),], broad_candidates[(broad_candidates$count!=0),])
    }
    sample_no <- ncol(seg_data_samples)
    b <- c(-1, -0.5, 0, 0.5, 1)
    y_axis_break <-c(-sample_no,-(sample_no/2) + -(sample_no/4), -(sample_no/2), -(sample_no/4),0, sample_no/4, sample_no/2,(sample_no/2) + (sample_no/4), sample_no)
    y_axis_label <- c("100","75", "50","25", "0","25", "50","75", "100" )
    
    cumFreq <- ggplot2::ggplot() + 
      ggplot2::geom_rect(data = candidates,  aes(xmin = seg_start, xmax = seg_end, ymax = count, ymin = 0 , fill = col))  +
      ggplot2::geom_vline(xintercept = genome_chr, colour = "grey") +
      ggplot2::geom_vline(xintercept = genome_centr, linetype="dotted", colour = "grey") +
      ggplot2::geom_hline( yintercept = 0, colour ="darkgrey") +
      ggplot2::scale_x_continuous(name = "Chromosome", breaks = axis_break, labels = axis_label, limits = c(0, 2881033286)) +
      ggplot2::theme_classic(base_size = 15) +
      ggplot2::guides(x = guide_axis(angle = 40)) +
      ggplot2::scale_y_continuous(name = "Occurence [%]", breaks = y_axis_break, labels = y_axis_label, limits = c(-sample_no, sample_no)) +
      ggplot2::scale_fill_discrete(guide="none") +
      ggplot2::scale_fill_manual(values = c("amplification" = colour.amplification,"deletion" = colour.loss), guide="none") +
      ggplot2::coord_cartesian(xlim = c(50001, 2881033286), expand = FALSE) +
      ggplot2::guides(x = guide_axis(n.dodge = 2))
  }
  return(cumFreq)
}
```

```{r}
setwd("/Volumes/PhD/CumulativeCopyNumberPaper/LUSC/capper_results")
files_multiseg_v1 <- list.files(path="/Volumes/PhD/CumulativeCopyNumberPaper/LUSC/capper_results/", pattern="multiseg_v1.csv", all.files=FALSE,full.names=FALSE)
files_multiseg_v2 <- list.files(path="/Volumes/PhD/CumulativeCopyNumberPaper/LUSC/capper_results/", pattern="multiseg_v2.csv", all.files=FALSE,full.names=FALSE)

files_singleseg_v1 <- list.files(path="/Volumes/PhD/CumulativeCopyNumberPaper/LUSC/capper_results/", pattern="singleseg_v1.csv", all.files=FALSE,full.names=FALSE)
files_singleseg_v2 <- list.files(path="/Volumes/PhD/CumulativeCopyNumberPaper/LUSC/capper_results/", pattern="singleseg_v2.csv", all.files=FALSE,full.names=FALSE)

graphdir <- "/Volumes/PhD/CumulativeCopyNumberPaper/LUSC/capper_results/graphs/"
filedir <- "/Volumes/PhD/CumulativeCopyNumberPaper/LUSC/capper_results/"

for(i in 1:length(files_singleseg_v1)){
x <- read.csv(paste(filedir,files_singleseg_v1[i], sep = ""), sep = " ")

x2 <- x
x2$segmean_OLD <- x2$segmean
  foreach::foreach(j = 1:nrow(x2)) %do% {
    if (x2$segmean_OLD[j] > (x2$noise[j] * thresh)) {
      x2$segmean[j] <- 1
    } else if (x2$segmean_OLD[j] < -(x2$noise[j] * thresh)) {
      x2$segmean[j] <- -1
    } else {
      x2$segmean[j] <- 0
    } 
  }
  
p1 <- singleFrequencyPlot(mSetsAnno, x2, colour.amplification = colour.amplification, colour.loss = colour.loss, thresh = thresh, array_type = "450k")
p2 <- overlayPlot(mSetsAnno, x, colour.amplification, colour.loss , array_type)

fileFreq = paste(tools::file_path_sans_ext(basename(files_singleseg_v1[i])), "_Freq.png", sep = "")
filepathFreq <- paste(graphdir,fileFreq, sep = "")

ggsave(filepathFreq, plot = p1, width = 6, height = 4, units = "in")

fileOverlay = paste(tools::file_path_sans_ext(basename(files_singleseg_v1[i])), "_Overlay.png", sep = "")
filepathOverlay <- paste(graphdir,fileOverlay, sep = "")

ggsave(filepathOverlay, plot = p2, width = 6, height = 4, units = "in")
}

for(i in 1:length(files_singleseg_v2)){
x <- read.csv(paste(filedir,files_singleseg_v2[i], sep = ""), sep = " ")
  
p1 <- singleFrequencyPlot(mSetsAnno, x, colour.amplification = colour.amplification, colour.loss = colour.loss, thresh = thresh, array_type = "450k")
p2 <- overlayPlot(mSetsAnno, x, colour.amplification, colour.loss , array_type)

fileFreq = paste(tools::file_path_sans_ext(basename(files_singleseg_v2[i])), "_Freq.png", sep = "")
filepathFreq <- paste(graphdir,fileFreq, sep = "")

ggsave(filepathFreq, plot = p1, width = 6, height = 4, units = "in")

fileOverlay = paste(tools::file_path_sans_ext(basename(files_singleseg_v2[i])), "_Overlay.png", sep = "")
filepathOverlay <- paste(graphdir,fileOverlay, sep = "")

ggsave(filepathOverlay, plot = p2, width = 6, height = 4, units = "in")
}

for(i in 1:length(files_multiseg_v1)){
x <- read.csv(paste(filedir,files_multiseg_v1[i], sep = ""), sep = " ")
p1 <- cumFreq(mSetsAnno, x, colour.amplification = colour.amplification, colour.loss = colour.loss, thresh = thresh, array_type = "450k")
p2 <- CCNV(mSetsAnno, seg_mpcf, target_ratios, array_type, colour.amplification, colour.loss, detail.regions = NULL, array_type)

fileFreq = paste(tools::file_path_sans_ext(basename(files_multiseg_v1[i])), "_Freq.png", sep = "")
filepathFreq <- paste(graphdir,fileFreq, sep = "")

ggsave(filepathFreq, plot = p1, width = 6, height = 4, units = "in")

fileOverlay = paste(tools::file_path_sans_ext(basename(files_multiseg_v1[i])), "_Overlay.png", sep = "")
filepathOverlay <- paste(graphdir,fileOverlay, sep = "")

ggsave(filepathOverlay, plot = p2, width = 6, height = 4, units = "in")
}

for(i in 1:length(files_multiseg_v2)){
x <- read.csv(paste(filedir,files_multiseg_v2[i], sep = ""), sep = " ")
p1 <- cumFreq(mSetsAnno, x, colour.amplification = colour.amplification, colour.loss = colour.loss, thresh = thresh, array_type = "450k")
p2 <- CCNV(mSetsAnno, seg_mpcf, target_ratios, array_type, colour.amplification, colour.loss, detail.regions = NULL, array_type)

fileFreq = paste(tools::file_path_sans_ext(basename(files_multiseg_v2[i])), "_Freq.png", sep = "")
filepathFreq <- paste(graphdir,fileFreq, sep = "")

ggsave(filepathFreq, plot = p1, width = 6, height = 4, units = "in")

fileOverlay = paste(tools::file_path_sans_ext(basename(files_multiseg_v2[i])), "_CCNV.png", sep = "")
filepathOverlay <- paste(graphdir,fileOverlay, sep = "")

ggsave(filepathOverlay, plot = p2, width = 6, height = 4, units = "in")
}
```

